### PR TITLE
[WIP] Improved async communication via lsp_server 

### PR DESCRIPTION
**Pull request by DavidDeSimone**
The purpose of this PR is to have a discussion and show the possibility of an architecture for speeding up IPC for use with an lsp server. 

The concept here is to parallelize as much of the process as possible. The core concept is that we perform the lisp object -> json object translation of the lisp thread, after which we hand off to the rust worker thread. The json object -> string translation happens on that thread, allowing for greater parallelization. 

While this code compiles, its more of a stub to simulate discussion then tested and ready to execute code. 

While not written out in this initial version, v2 of this concept would be that ALL IPC (both reads and writes to the LSP server) happens on the rust worker, and the resultant json is only parsed once by our json library, and then the appropriate lisp callbacks are dispatched. This will mean that instead of calling json_decode (or any alternative) on the result string from the server, we call our parse method in rust, transfer that json object to the lisp thread, dispatch the callbacks while allocating lisp objects as necessary.

### CODE REVIEW

File: rust_src/src/parsing.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    };
+
+    let data = UserData::new(jsonm);
+    crate::ng_async::Fasync_send_message(pipe, data.into())
+}
+
+// Option 1: Write functions that look something like whats defined below
+// Option 2: Compile emacs-ng with json enabled, and use the
+// emacs functions defined in json.c, pass json_t* around instead of
+// serde_json::Value objects
```
Comment (brotzeit) [inline rust_src/src/parsing.rs:41]: I'd prefer using serde_json or do you think it's better to use json.c ?

**Comment by brotzeit** [comment]
@yyoncho what are your thoughts on this ? I think we should also keep in mind how the implementation can be used in lsp-clients like lsp-mode.

**Comment by yyoncho** [comment]
@DavidDeSimone did you had a chance to take a look at the https://github.com/rust-analyzer/lsp-server ?

Here is my playground, which can be used for some pieces of code.

https://github.com/emacs-ng/emacs-ng/blob/1b262c8b9de6ee0cfdd5bc4bf6a5254aec2db7a1/rust_src/src/fastjsonrpc.rs

The question here is whether we could use its channel model or better just to strip the parsing logic? The other open question is whether to have one reading thread for each jsonrpc connection(maybe we should). To me, it seems like the channels won't be needed since it will introduce one redundant transfer unless we want to move more logic out of the IO reading thread. 

About this POC: generally, the lsp-mode issue with jsonrpc is with receiving and parsing. Most of the time we are sending very small messages for which we will be good even if we do everything on the main thread. Of course, it will be good if we optimize it as well but IMHO we should focus on the receiving pipeline.

**Comment by DavidDeSimone** [comment]
@yyoncho 

Thank you for the feedback. I will take a deeper look at https://github.com/rust-analyzer/lsp-server and your playground link. 

My intention is to review the above and update this POC with my thoughts once I've had a chance to think through what was posted above. I will likely shift to a stronger focus on the receiving and parsing side based on your feedback.

**Comment by DavidDeSimone** [comment]
@yyoncho I've updated this PR based on your feedback, and the code in your playground (https://github.com/emacs-ng/emacs-ng/blob/1b262c8b9de6ee0cfdd5bc4bf6a5254aec2db7a1/rust_src/src/fastjsonrpc.rs)

I have opt'd to not use crossbeam, and instead use the "EmacsPipe" struct that was defined in earlier PRs, however I left a note about some options we have on its usage in the source code.

**Comment by yyoncho** [comment]
Looks good to me. It looks like we have pretty much all needed to do the whole end-to-end testing and benchmarking. 

For sending elisp structures, the simplest solution to unblock the testing is to pass Elisp string -> (convert it to rust string) -> write it into stdin. 

Now, for receiving, we have two options:

1. Converting the Message to emacs data structure in the elisp thread. This is simpler and easier to get it right but we need a benchmark. 
2. Wrap the Message in UserData and provide operation on it which is like this:

```
(get-property user-data key)
```
which may return:

* Elisp primitive (string, bool, etc)
* Vector of elisp primitive 
* Another user-data
* Vector of user-data

This will allow us to lazily convert the message to elisp structure when consumed. 

The challenge is that we want `(get-property USER-DATA "foo")` when called twice to return the same object. This means that we have to cache them in the UD and somehow prevent GC. I have filed https://debbugs.gnu.org/cgi/bugreport.cgi?bug=42032 and then I thought that we may fix everything if we keep elisp map which is `ud` -> `list of cache objects`. Then when the finalizer of UD is called we will clean that map. But I am not sure if this is going to work and perform well.

**Comment by yyoncho** [comment]
> 1. Converting the Message to emacs data structure in the elisp thread. This is simpler and easier to get it right but we need a benchmark.

I suspect the slowest thing to convert will be the strings. You mentioned somewhere that we might somehow do that in the bkg thread. If we can do that, then this might be the preferred solution. 

One more point to mention: option 2. might result in less elisp objects and less GC which is also an issue.

**Comment by DavidDeSimone** [comment]
I am going to remove the RFC from this PR and transition it to a PR. Priority is to transition this to a basic implementation and begin benchmarking. 

> The challenge is that we want (get-property USER-DATA "foo") when called twice to return the same object. This means that we have to cache them in the UD and somehow prevent GC. I have filed https://debbugs.gnu.org/cgi/bugreport.cgi?bug=42032 and then I thought that we may fix everything if we keep elisp map which is ud -> list of cache objects. Then when the finalizer of UD is called we will clean that map. But I am not sure if this is going to work and perform well.

I have a few thoughts on this issue, however I will need to benchmark some of the possible solutions. I think that this aspect of the project has the potential to cause the biggest performance issues if we don't handle it correctly. I will code up some sample tests and post results.

File: src/data.c [hunk 2]
```diff
@@ -3788,6 +3788,8 @@ syms_of_data (void)
 {
   Lisp_Object error_tail, arith_tail;
   rust_init_syms();
+  DEFSYM (QCnull, ":null");
+  DEFSYM (QCfalse, ":false");
```
Comment (DavidDeSimone) [inline src/data.c:3792]: I can remove this code and add calls to `def_lisp_sym!` once I bring that over from the remacs codebase.

**Comment by yyoncho** [comment]
That fails - https://gist.github.com/yyoncho/2019e3b9be75f4c6dde15a6861abe0ab

**Review by yyoncho — COMMENTED** [review]
What about merging this to make https://github.com/emacs-ng/emacs-ng/pull/33 clearer?

**Comment by DavidDeSimone** [comment]
I'm fine with merging this as-is, do you agree @brotzeit ?

File: rust_src/src/parsing.rs [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+    let mut plist = unsafe { Fprocess_plist(proc) };
+    plist = unsafe { Fplist_put(plist, QCjson_config, user_ptr) };
+    unsafe { Fset_process_plist(proc, plist) };
+
+    true
+}
+
+/*
+
+#[lisp_fn]
```
Comment (harryfei) [inline rust_src/src/parsing.rs:166]: Dirty code

File: rust_src/Cargo.toml [hunk 4]
```diff
@@ -27,8 +28,10 @@ itertools = "0.8"
 lazy_static = "1.2"
 libc = "0.2"
 line-wrap = "0.1.1"
+lsp-server = "0.5.0"
```
Comment (harryfei) [inline rust_src/Cargo.toml:31]: Do we actually use this crate?

**Pull request merged**